bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/FileInfo.java

```diff
@@ -549,6 +549,7 @@ public synchronized void moveToNewLocation(File newFile, long size) throws IOExc
         }
         fc = new RandomAccessFile(newFile, mode).getChannel();
         lf = newFile;
+        deleted = false;
```

Comment:
it'd be better to have a method `delete(boolean markAsDeleted)` and passing false in this method

in your new code we still have few lines where the `deleted` flag is equals to `true` and it can lead in undefined behaviour

Reply:
Seems that all access methods of `deleted` and `fc` are `synchronized`. So I think it is safe ?

Actually, I could keep `deleted` untouched by using `lf.delete()` directly to delete file instead of `FileInfo.delete` which is used as spying point to mimic crashing behavior in tests. But, this way I have to introduce an extra spying point to intercept. Should I go this way and make spying point more explicitly ?

Reply:
@nicoloboschi are you ok with this?

Reply:
@nicoloboschi ping

Reply:
> `moveToNewLocation` is not expected the be in deleted state at all during the rename


It is only correct in ideal situation. Could this operation happen cross filesystems or even devices ? I think this method make no assumption about this. From my side, `delete` in `moveToNewLocation` is more like a conservative operation than solely `Files.delete()`(which throw exception on deletion failure). After failure of `FileInfo.delete`, `FileInfo` is unusable as `FileInfo.fc` is closed and `FileInfo.lf` is not guaranteed to exist.

I have no preference from approaches:
* Current: `FileInfo.delete()` and throws exception on failure.
* Alternative: `FileInfo.lf.delete()` and mark `deleted` and throws exception on failure.

I think they are identical.

Reply:
I have reverted related changes since #3421 fixed this.